controlplane/kubeadm/controllers/scale.go

```diff
@@ -146,3 +132,11 @@ func (r *KubeadmControlPlaneReconciler) scaleDownControlPlane(
 	// Requeue the control plane, in case there are additional operations to perform
 	return ctrl.Result{Requeue: true}, nil
 }
+
+func selectMachineForScaleDown(controlPlane *internal.ControlPlane) (*clusterv1.Machine, error) {
+	machines := controlPlane.Machines
+	if needingUpgrade := controlPlane.MachinesNeedingUpgrade(); needingUpgrade.Len() > 0 {
+		machines = needingUpgrade
+	}
+	return controlPlane.MachineInFailureDomainWithMostMachines(machines)
```

Comment:
If I'm not wrong, in case of upgrades, this is considering only the machines to upgrades, not all the control-plane machines.
Is it possible this leads to an un-unexpected machine distribution? probably this is compensated by scale-up always looking at all the machines for placement...

Reply:
I don't think it will change the machine distribution. We scale down the FD with the most un-upgraded machines, but we scale back up by smallest FD overall

Here's a run-through of what I understand the algorithm to be: https://gist.github.com/benmoss/9914fb4e09e1e4fed4a651119e983298